spd/utils/target_ci_solutions.py

```diff
@@ -0,0 +1,421 @@
+        """Permute to show identity pattern."""
+        if self.method == "hungarian" or (self.method == "auto" and min(ci_array.shape) < 500):
+            return permute_to_identity_hungarian(ci_array)
+        else:
+            return permute_to_identity_greedy(ci_array)
+
+
+class DenseCIPattern(TargetCIPattern):
+    """Dense columns pattern: exactly K components should be active.
+
+    Expects sparsity where exactly K columns (components) have non-zero entries.
+
+    Args:
+        k: Number of columns that should be active
+        min_entries: Minimum number of strong activations (> 1-tolerance)
+                    required for a column to be considered "active"
+
+    Error computation:
+        - For first k columns: count how many are missing required strong activations
+        - For remaining columns: count any activations (should be completely inactive)
+        - Total error = active_column_deficits + inactive_column_violations
+    """
+
+    def __init__(self, k: int, min_entries: int = 1):
+        self.k = k
+        self.min_entries = min_entries
+
+    @override
+    def _verify_inputs(self, ci_array: Float[Tensor, "batch C"]) -> None:
+        super()._verify_inputs(ci_array)
+        _, c = ci_array.shape
+        if c < self.k:
+            raise ValueError(f"Expected at least {self.k} columns, got {c}")
+
+    @override
+    def distance_from(self, ci_array: Float[Tensor, "batch C"], tolerance: float = 0.1) -> int:
+        self._verify_inputs(ci_array)
+        sorted_ci = permute_to_dense(ci_array)[0]
+
+        strong_activations_per_column = (sorted_ci >= 1 - tolerance).sum(dim=0)
+        missing_strong_activations = torch.clamp(
+            self.min_entries - strong_activations_per_column, min=0
+        )
+        first_k_column_error = missing_strong_activations[: self.k].sum().item()
+
+        weak_activations_per_column = (sorted_ci > tolerance).sum(dim=0)
+        inactive_column_error = weak_activations_per_column[self.k :].sum().item()
+
+        return int(first_k_column_error + inactive_column_error)
+
+    @override
+    def permute_for_display(
+        self, ci_array: Float[Tensor, "batch C"]
+    ) -> tuple[Float[Tensor, "batch C"], Int[Tensor, " C"]]:
+        """Permute to show dense columns pattern."""
+        return permute_to_dense(ci_array)
+
+
+class TargetCISolution:
+    """Collection of expected patterns for different modules in a model.
+
+    The module_targets dictionary can use fnmatch-style patterns as keys:
+    - Explicit module names: {"layers.0.mlp_in": IdentityCIPattern(...)}
+    - Wildcard patterns: {"layers.*.mlp_in": IdentityCIPattern(...)}
+    - Mixed patterns: {"layers.*.mlp_*": pattern, "specific.module": pattern}
+
+    Patterns are expanded at runtime when given actual module names.
+    First matching pattern wins for each module name.
+    """
+
+    def __init__(
+        self, module_targets: dict[str, TargetCIPattern], expected_matches: int | None = None
```

Comment:
I think the expected_matches argument is overkill. I'd remove it everywhere. Especially for an eval. If a user screws up with their patterns, they'll work it out pretty quickly.